### DISCUSSION TITLE
Systemd socket-based activation for http server?

### DISCUSSION THREAD

**Discussion by cryobry** [Motion-Project/motion#1481] [General]
Would it be possible to add a systemd .socket activated service file to the default installation (when using the httpd server)? This would allow my server to use a lot less resources when the stream is not being accessed via a web interface.

**Comment by tosiara** [comment]
I'm not familiar with "systemd .socket activated service file". Can you provide more details or maybe an example of other programs using it?

**Comment by cryobry** [comment]
https://www.freedesktop.org/software/systemd/man/systemd.socket.html
    
    # /usr/lib/systemd/system/cockpit.socket
    [Unit]
    Description=Cockpit Web Service Socket
    Documentation=man:cockpit-ws(8)
    Wants=cockpit-motd.service

    [Socket]
    ListenStream=9090
    ExecStartPost=-/usr/share/cockpit/motd/update-motd '' localhost
    ExecStartPost=-/bin/ln -snf active.motd /run/cockpit/motd
    ExecStopPost=-/bin/ln -snf /usr/share/cockpit/motd/inactive.motd /run/cockpit/motd

    [Install]
    WantedBy=sockets.target`

The socket watches for access to the ListenStream port, after which it executes the Exec lines. This could be used to start/stop the motion service on-demand.

**Comment by tim-seoss** [comment]
This implies (I believe) that motion would not usually run - it would be started on demand by systemd when a client (e.g. a web browser) connects, and would exit after the streaming connection is closed.  It wouldn't be able to do it's usual motion detection tasks etc.

If that's the case, then I think it might be a better idea to just implement this with a simple gstreamer pipeline or similar, rather than add this to motion?

If I've got this wrong, and there's a use-case for this which would utilise motion's major features, then could you outline what sort of usage you were thinking of?

**Comment by cryobry** [comment]
On point one you are correct. This is part of a lab server setup with a
webcam that monitors a robotic arm. 99% of the time the server is doing
heavy calculations and needs as many resources available as possible, but
occasionally we need to view the robot's actions remotely in case a problem
arises.

On point two I am unfamiliar with gstreamer pipelines. I know this can work
via systemd sockets, although there may certainly be a better way as you
have indicated.
